### PR TITLE
Create the `expansions` table and models using STI

### PR DESCRIPTION
**Pull request by WilliamMcCumstie**
This is a smallish PR that mainly focuses on updating the database to include `expansions`. Each `Expansion` includes an `ExpansionType`, that is stored in it's own table (see: https://github.com/alces-software/alces-flight-center/commit/53903a2976c791fa885363c8f70a54b49cd25173).

Each expansion is either a `ComponentExp..` or `DefaultExp..` expansion. Currently only the associations exist which will expanded on in feature `PR`s. The idea being that a `DefaultExp..` will be used to create a `ComponentExp` when a new `Component` is created (unless otherwise specified).

Another commit is still required to fix `rspec` but further discussion is required.

### CODE REVIEW

**Comment by bobwhitelock** [comment]
Rebased on `develop` so CI should work properly.

File: spec/factories.rb [hunk 1]
```diff
@@ -1,5 +1,16 @@
...
+    slot "my-expansion"
+    ports 4
+    expasion_type expansion_type
+    type "expansion"
+  end
+
+  factory :expansion_type do
+    name "ExpansionDummy"
+  end
+
```
Comment (bobwhitelock) [inline spec/factories.rb]: I don't think we need these factories at the moment: they're unused and so there's also potential for them to get out-of-sync with the corresponding models, and the flipside of this is we don't want to expend effort maintaining unused code.

Also (not that this matters much since we don't need them), there's a typo in the `expasion_type` field - this should be `expansion_type`, and since this relation refers to the model of the same name you should only need to specify it once.

File: spec/models/application_record_spec.rb [hunk 2]
```diff
@@ -20,14 +20,19 @@ def mock_request_user
...
+
       ActiveRecord::Base.connection.tables.each do |table|
         begin
           klass = table.singularize.camelize.constantize
 
-          # Users are special, they have a relation with a Site but are also
-          # globally available, i.e. able to be read by any other User.
-          next if klass == User
+          # Irregular models:
+          # Users: have a relation with a Site but are also global
```
Comment (bobwhitelock) [inline spec/models/application_record_spec.rb]: Can we retain the full explanation for why `User`s are an exception (from the left) here please?

File: spec/models/application_record_spec.rb [hunk 3]
```diff
@@ -20,14 +20,19 @@ def mock_request_user
...
+          # Exception: is a base class for STI and is neither
```
Comment (bobwhitelock) [inline spec/models/application_record_spec.rb]: I think this should say `Expansion` rather than `Exception`. Also the 'and is neither' part doesn't make much sense, can you rephrase what you mean?

File: spec/models/expansion_type_spec.rb [hunk 4]
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
+
+RSpec.describe ExpansionType, type: :model do
+  pending "add some examples to (or delete) #{__FILE__}"
```
Comment (bobwhitelock) [inline spec/models/expansion_type_spec.rb]: Can we follow the suggestion made here and delete these files so they don't spam the logs?

File: app/models/expansion.rb [hunk 5]
```diff
@@ -0,0 +1,3 @@
+class Expansion < ApplicationRecord
+  belongs_to :expansion_type
```
Comment (bobwhitelock) [inline app/models/expansion.rb]: Validations for the presence of `slot`, `ports`, and maybe `type` (to avoid creating an `Expansion` which isn't a specific subclass) would be useful here.

File: spec/models/application_record_spec.rb (hunk 2 above)
Comment (bobwhitelock) [reply spec/models/application_record_spec.rb]: Don't worry about this comment actually, I'm tweaking how this test generation works and will make this change in the process.

File: spec/models/application_record_spec.rb (hunk 3 above)
Comment (bobwhitelock) [reply spec/models/application_record_spec.rb]: As above, don't worry about this comment.

File: spec/factories.rb (hunk 1 above)
Comment (WilliamMcCumstie) [reply spec/factories.rb]: The factories have been removed in: https://github.com/alces-software/alces-flight-center/pull/30/commits/08df57611d9eecb6383d1ad621ee7261d8d31698

I have came across that typo else, where it should be fixed now?

File: app/models/expansion.rb (hunk 5 above)
Comment (WilliamMcCumstie) [reply app/models/expansion.rb]: The type validation has been added in: https://github.com/alces-software/alces-flight-center/pull/30/commits/f89e934282edebe62a7b449f7f541d13266ce455.

Will do the others soon

File: spec/models/application_record_spec.rb (hunk 3 above)
Comment (WilliamMcCumstie) [reply spec/models/application_record_spec.rb]: Ooops, be you to it: https://github.com/alces-software/alces-flight-center/pull/30/commits/38a9b51007fd3796e947b164f5f9d38822ef102a

File: app/models/expansion.rb (hunk 5 above)
Comment (WilliamMcCumstie) [reply app/models/expansion.rb]: Fixed in: https://github.com/alces-software/alces-flight-center/pull/30/commits/b58f4aca2194bd6fdede4606a8b31d792b81efb9

**Comment by WilliamMcCumstie** [comment]
Ok @bobwhitelock, I have made those changes. Should be ready to merge unless you have got something else?

**Comment by bobwhitelock** [comment]
Nope looks great, thanks!
